File: src/SyntaxDemo.java
import java.util.ArrayList;

public class SyntaxDemo {
    public static void main(String[] args) {
        System.out.println(foo1("Hello world!"));
        System.out.println(greeting("Josh Hug"));
        System.out.println(getMessage(404));
    }

    private static int foo1(String argument1) {
        int length;
        length = argument1.length(); // this *should* be final, but no one does that
        int sum = 0;
        for (int i = 0; i < length; i += 1) {
            sum += argument1.charAt(i); // "strongly typed" lmao
        }
        return sum;
    }

    // can get super tedious with lots of little string insertions like this
    private static String greeting(String name) {
        return "Hello, " + name + "!";
    }

    // switch statements are verbose
    private static String getMessage(int responseCode) {
        String message;
        switch (responseCode) {
            case 200:
            case 201:
                message = "u good";
                break;
            case 404:
                message = "page not found";
                break;
            case 500:
                message = "rip";
                break;
            default:
                message = "Unexpected error";
                break;
        }
        return "Response " + responseCode + ": " + message;
    }

    private static ArrayList<String> bar() {
        ArrayList<String> strings = new ArrayList<>();
        strings.add("one");
        strings.add("two");
        strings.add("three");
        return strings;
    }

    private static void printBar() {
        ArrayList<String> strings = bar();

        // someone else can mess with your list!
        strings.clear();
        strings.add("four");
        System.out.println(strings.get(0));
    }
}

File: src/SyntaxDemo.kt
fun main() {
    println(foo1("Hello world!"))
    println(foo1A("Hello world!"))
    println(foo1MAX("Hello world!"))
    println(greeting("Josh Hug"))
    println(getMessage(404))
}

// no semicolons!
fun foo1(argument1: String): Int {
    var length: Int
    length = argument1.length // no primitives
    var sum: Int = 0
    for (i in 0 until length) { // loops in ranges (like python)
        sum += argument1[i].toInt() // no implicit casts
    }
    return sum
}

// type inference and immutable vals
fun foo1A(argument1: String): Int {
    val length = argument1.length // we'll never change this, so why var?
    var sum = 0 // 0 is an int, so sum's an int too
    for (i in 0 until length) {
        sum += argument1[i].toInt()
    }
    return sum
}

// max kotlin!!
fun foo1MAX(argument1: String) = argument1.chars().sum()

// string placeholders
fun greeting(name: String): String {
    return "Hello, $name!"
}

// when statements are inlineable, succint, no breaks
fun getMessage(responseCode: Int): String {
    // didn't have to use a mutable variable!
    val message = when (responseCode) {
        200, 201 -> "u good"
        404 -> "page not found"
        500 -> "rip"
        else -> "Unexpected error"
    }
    return "Response $responseCode: $message"
}

fun bar(): List<String> {
    return listOf("one", "two", "three")
}

fun barMutable(): MutableList<String> {
    return mutableListOf("one", "two", "three")
}

fun printBar() { // no return type btw
//    this won't compile! can't remove, add, or replace anything in immutable list
//    val strings = bar()
//    strings.clear()
//    strings.add("four")

    val strings = barMutable()
    strings.clear()
    strings.add("four")
    println(strings[0])
}
